pkg/controller/lifecycle/actuator.go

```diff
@@ -0,0 +1,245 @@
+	client         client.Client
+	config         *rest.Config
+	decoder        runtime.Decoder
+	extensionClass extensionsv1alpha1.ExtensionClass
+
+	serviceConfig config.Configuration
+
+	logger logr.Logger
+}
+
+// Reconcile the Extension resource.
+func (a *actuator) Reconcile(ctx context.Context, log logr.Logger, ex *extensionsv1alpha1.Extension) error {
+	namespace := ex.GetNamespace()
+	internal := !gutil.IsShootNamespace(namespace)
+
+	var cluster *extensions.Cluster
+	var err error
+	if !internal {
+		cluster, err = controller.GetCluster(ctx, a.client, namespace)
+		if err != nil {
+			return err
+		}
+	}
+
+	certConfig := &service.CertConfig{}
+	if ex.Spec.ProviderConfig != nil {
+		if _, _, err := a.decoder.Decode(ex.Spec.ProviderConfig.Raw, nil, certConfig); err != nil {
+			return fmt.Errorf("failed to decode provider config: %+v", err)
+		}
+		if errs := validation.ValidateCertConfig(certConfig, cluster); len(errs) > 0 {
+			return errs.ToAggregate()
+		}
+	}
+
+	values, err := a.createValues(ctx, certConfig, cluster, namespace)
+	if err != nil {
+		return err
+	}
+
+	if !internal && !controller.IsHibernated(cluster) {
+		if err := a.createShootResources(ctx, *values); err != nil {
+			return err
+		}
+		if err := a.createSeedResources(ctx, *values); err != nil {
+			return err
+		}
+	} else {
+		if err := a.createInternalResources(ctx, *values); err != nil {
+			return err
+		}
+	}
+
+	return a.updateStatus(ctx, ex, certConfig)
+}
+
+// Delete the Extension resource.
+func (a *actuator) Delete(ctx context.Context, log logr.Logger, ex *extensionsv1alpha1.Extension) error {
+	namespace := ex.GetNamespace()
+	internal := !gutil.IsShootNamespace(namespace)
+
+	a.logger.Info("Component is being deleted", "component", "cert-management", "namespace", namespace)
+	if !internal {
+		if err := a.deleteShootResources(ctx, namespace); err != nil {
+			return err
+		}
+		return a.deleteSeedResources(ctx, namespace)
+	}
+
+	return a.deleteInternalResources(ctx, namespace)
+}
+
+// ForceDelete the Extension resource.
+func (a *actuator) ForceDelete(ctx context.Context, log logr.Logger, ex *extensionsv1alpha1.Extension) error {
+	return a.Delete(ctx, log, ex)
+}
+
+// Restore the Extension resource.
+func (a *actuator) Restore(ctx context.Context, log logr.Logger, ex *extensionsv1alpha1.Extension) error {
+	return a.Reconcile(ctx, log, ex)
+}
+
+// Migrate the Extension resource.
+func (a *actuator) Migrate(ctx context.Context, log logr.Logger, ex *extensionsv1alpha1.Extension) error {
+	// Keep objects for shoot managed resources so that they are not deleted from the shoot during the migration
+	if err := managedresources.SetKeepObjects(ctx, a.client, ex.GetNamespace(), v1alpha1.CertManagementResourceNameShoot, true); err != nil {
+		return err
+	}
+
+	return a.Delete(ctx, log, ex)
+}
+
+func (a *actuator) createValues(ctx context.Context, certConfig *service.CertConfig, cluster *controller.Cluster, namespace string) (*Values, error) {
```

Comment:
Nit: Pass `InternalDeployment` (or renamed version - see other comment about `internal`) to avoid using `gutil.IsShootNamespace` another time.